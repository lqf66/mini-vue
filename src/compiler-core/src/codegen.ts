import { NodeTypes } from "./ast"
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context

  if (ast.helpers.length > 0) {
    genFunctionPreamble(ast, context)
  }

  push('return ')
  const functionName = 'render'
  const args = ['_ctx, _cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}) {`)

  push(' return ')
  genNode(ast.codegenNode, context)
  push(' }')

  return {
    code: context.code
  }
}

function genFunctionPreamble(ast: any, context) {
  const { push } = context
  const VueBinging = 'vue'
  const aliasHelpers = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
  push(`const { ${ast.helpers.map(aliasHelpers).join(', ')} } = ${VueBinging} `)
  push('\n')
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
    default:
      break
  }
}

function genText(node: any, context: any) {
  const { push } = context
  push(`'${node.content}'`)
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(`)`)
}

function genExpression(node: any, context: any) {
  const { push } = context
  push(`${node.content}`)
}
