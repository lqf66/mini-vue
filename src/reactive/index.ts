export { ref, proxyRefs } from './ref'
export { effect } from './effect'
export { readonly, shallowReadonly } from './reactive'