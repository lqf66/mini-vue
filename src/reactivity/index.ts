/*
 * @Author: Lqf
 * @Date: 2021-12-19 12:31:40
 * @LastEditors: Lqf
 * @LastEditTime: 2021-12-22 21:24:37
 * @Description: 我添加了修改
 */

export function add(a, b) {
  return a + b
}