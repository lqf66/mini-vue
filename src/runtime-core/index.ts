/*
 * @Author: Lqf
 * @Date: 2021-12-25 19:24:59
 * @LastEditors: Lqf
 * @LastEditTime: 2021-12-26 15:38:11
 * @Description: 我添加了修改
 */


export { createApp } from "./createApp"
export { h } from './h'