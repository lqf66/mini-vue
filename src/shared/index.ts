/*
 * @Author: Lqf
 * @Date: 2021-12-19 16:51:15
 * @LastEditors: Lqf
 * @LastEditTime: 2021-12-27 19:52:02
 * @Description: 我添加了修改
 */

export const extend = Object.assign

export const isObject = (val) => {
  return val !== null && typeof val === 'object'
}

export const hasChanged = (value, newValue) => {
  return !Object.is(value, newValue)
}

export const hasOwn = (val, key) => Object.prototype.hasOwnProperty.call(val, key)
