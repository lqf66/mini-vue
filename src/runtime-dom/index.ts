/*
 * @Author: Lqf
 * @Date: 2022-01-01 15:06:32
 * @LastEditors: Lqf
 * @LastEditTime: 2022-01-01 15:58:59
 * @Description: 我添加了修改
 */

import { createRenderer } from "../runtime-core"

function createElement(type) {
  console.log('----createElement----')

  return document.createElement(type)
}

function patchProp(el, key, val) {
  console.log('----patchProp----')
  const isOn = (key: string) => /^on[A-Z]/.test(key)
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase()
    el.addEventListener(event, val)
  } else {
    el.setAttribute(key, val)
  }
}

function insert(el, container) {
  console.log('----insert----')
  container.append(el)
}

const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert
})

export function createApp(...args) {
  return renderer.createApp(...args)
}


export * from '../runtime-core'
