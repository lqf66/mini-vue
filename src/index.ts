export * from './runtime-dom'

import { baseCompile } from './compiler-core/src'
import * as runtimeDom from './runtime-dom'
import { registerRuntimeCompiler } from './runtime-dom'

function compileToFunction(template) {
  const { code } = baseCompile(template)

  const render = new Function('Vue', code)(runtimeDom)
  console.log('render: ', render)

  return render
}

registerRuntimeCompiler(compileToFunction)
