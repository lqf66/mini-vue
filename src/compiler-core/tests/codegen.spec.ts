import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformExpress } from "../src/transforms/transformExpression"

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformExpress]
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
