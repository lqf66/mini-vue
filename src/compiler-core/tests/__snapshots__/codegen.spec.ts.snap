// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { toDisplayString: _toDisplayString, create-element-vnode: _create-element-vnode } = vue 
return function render(_ctx, _cache) { return _create-element-vnode(\\"div\\", null, 'hi' + _toDisplayString(_ctx.message)) }"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = vue 
return function render(_ctx, _cache) { return _toDisplayString(_ctx.message) }"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache) { return 'hi' }"`;
